tests/integration/panic.rs

```diff
@@ -0,0 +1,196 @@
+use chalk_integration::interner::{ChalkIr, RawId};
+use chalk_ir::*;
+use chalk_rust_ir::*;
+use chalk_solve::{RustIrDatabase, SolverChoice};
+use std::sync::Arc;
+
+#[derive(Debug, Default)]
+struct MockDatabase {
+    panic: bool,
+}
+
+#[allow(unused_variables)]
+impl RustIrDatabase<ChalkIr> for MockDatabase {
+    fn custom_clauses(&self) -> Vec<ProgramClause<ChalkIr>> {
+        if self.panic {
```

Comment:
So, I think ideally we want this impl to panic in each of the functions to *really* test for panic safety more thoroughly. Do you want to try to do this?

I imagine you could make `panic` be an `Option<enum>` with variants `CustomClauses`, `TraitDatum`, etc. Where if `panic` is that variant when the function is called, then that function panics. I think if the solver panic-safe enough to solve the test goal given here, that's a great start.

Reply:
I little more work, but I think I can do that.

The problem that I see is that we actually basing on internal implementation anyway - because most of functions are `unimplemented!()` (so they just panic). The problem is that to really test all the functions I would need some program and goal which actually call *every* single DB function, which I think can take me something around eternity with my current Chalk knowledge (ok, maybe slightly overestimated).

Reply:
Haha yeah.

I'm thinking. It might be worth going through and figuring out *when* in the `chalk-engine` lifecycle the db methods are called. I imagine *most* of them are called when building the table. (In which case it probably isn't helpful to add extra tests for each method)

Reply:
I was thinking on this a little. As you said - those methods are called mostly while building a table. This is however possible, that some method might be called somewhere else (maybe is now, or maybe will be in week, month, year, whatever). The problem is, that we would like to test the context on which method panicked, not the panicking method itself. So if method `impls_for_trait` is called for specific program 4 times, we would actually need to panic on it only on first call, only on second call, and so on. This would literally take an eternity, and I can do it for any timeframe you will tell me to do this (I mean I am able too, not willing to - I would like to help, not to die out of boredom).

What it comes to, that to really test all cases, we need to test all flows, and this is like at least big problem. My opinion is, that proper way to approach this is maybe create some fuzz tests, which are creating bigger and bigger programs and goals, and then try to solve the goal (unwinding panic), and then try to solve another goal on this program (also complicated, but such we already tested does not panic). This way we may find some actual flow which is a candidate to be improved.

However proposed solution is not easy, and I am not raising my hand to do this - I don't think how to actually approach such kind of testing.